File: matchsticks/C++/matchsticks.cpp
#include <iostream>
using namespace std;

bool c = true;
int s = 21;

void takeOrEnd(int t) {
    if (s - t <= 0) {
        cout << "game over" << endl;
        cout <<  "the winner is " +  (string) (c ? "you" : "computer") << endl;
        s = 0;
    } else {
        s -= t;
    }
}

void printSticks() {
  cout << "there are " + to_string(s) +  " matchstick(s) on the table" << endl;
   cout << string(s, '|') <<endl;
}

int  main() {

    cout << "there are 21 matchsticks on the table" << endl;
    cout << "computer's first"<< endl;

    while (s != 0) {
        if (c) {
            cout << "computer's thinking" << endl;
            int r = (s - 1) % 4;
            int t = (r == 0) ? 1 : r;

            cout << "computer takes $t" << endl;
            takeOrEnd(t);
            c = !c;

        } else {
            cout << "it is your turn" << endl;
            int t;

            do {
                cout << "you can take 1, 2 or 3!" << endl;
                cout <<  "so how many sticks will you take?" << endl;
                cin >> t;
            } while (!(t == 1 || t == 2 || t == 3));

            takeOrEnd(t);
            c = !c;
        }

        printSticks();
    }
    return 0;
}

File: matchsticks/JavaScript/matchsticks.js
var c = true
var s = 21

function main(args) {

    console.log("there are 21 matchsticks on the table");
    console.log("computer's first");

    while (s != 0) {
        if (c) {
            console.log("computer's thinking")
            var r = (s - 1) % 4
            var t = (r == 0) ? 1 : r

            console.log("computer takes $t")
            takeOrEnd(t)
            c = !c

        } else {
            console.log("it is your turn")
            var t

            do {
                console.log("you can take 1, 2 or 3!")
                t = prompt("so how many sticks will you take?")
            } while (!(t == 1 || t == 2 || t == 3))

            takeOrEnd(t)
            c = !c
        }

        printSticks()
    }
}

function takeOrEnd(t) {
    if (s - t <= 0) {
        console.log("game over")
        console.log("the winner is " + ( (c) ? "you" : "computer"))
        s = 0
    } else {
        s -= t
    }
}

function printSticks() {
    console.log(`there are ${s} matchstick(s) on the table`)
    console.log("|".repeat(s))
}